### ISSUE TITLE
SIGSEGV during network interruption

### ISSUE THREAD

**Issue by doktorstick** [sahlberg/libnfs#185]
The base for my libnfs is 1.10. You'll note that the line numbers aren't the same. It has a an `rpc->is_conneced` patch from you around the July 2016 timeframe. Also, I've added a set of `timeout` APIs that is used in the `wait_for_*` routines (you'll see the `timeout` field in the `rpc` struct below). This was to address a problem where libnfs would spin for forever on a server that quietly went away. (Still testing before I push it upstream.)

Anyway, what I observed is the `fsync` operation aborts and I get a crash in `LIBNFS_LIST_REMOVE`. I've not started digging into the the list management, yet; I wanted to get this bug posted in case it's a known/fixed or a eureka! moment.

```
(gdb) frame 1
#1  0x00007fffeadbdf6b in nfs_destroy_context (nfs=0x7fffe0002af0) at libnfs.c:475
475        rpc_destroy_context(nfs->rpc);
(gdb) p *nfs
$6 = {rpc = 0x7fffe0002b50, server = 0x7fffe010b200 "10.151.3.161", export = 0x7fffe01081a0 "/ifs/export/clump", rootfh = {data = {data_len = 32,
      data_val = 0x7fffe010b220 "\213k\376\273\004"}}, readmax = 1048576, writemax = 1048576, cwd = 0x7fffe0000c40 "/", dircache = 0x0, mask = 18,
  auto_traverse_mounts = 0, nested_mounts = 0x0}
(gdb) frame 0
#0  0x00007fffeadbd071 in rpc_destroy_context (rpc=0x7fffe0002b50) at init.c:296
296                LIBNFS_LIST_REMOVE(&q->head, pdu);
(gdb) p *rpc
$7 = {magic = 3336856629, fd = 125, old_fd = 0, is_connected = 1, error_string = 0x7fffe010a000 "fsync call failed with \"Command was cancelled\"",
  connect_cb = 0x0, connect_data = 0x7fffe0107f90, auth = 0x7fffe0107d20, xid = 3868610387, encodebuf = 0x7fffe0006d10 "\200", encodebuflen = 1052672,
  outqueue = {head = 0x0, tail = 0x0}, udp_src = {ss_family = 0, __ss_align = 0, __ss_padding = '\000' <repeats 111 times>}, waitpdu = {{head = 0x0,
      tail = 0x0} <repeats 398 times>, {head = 0x7fffe0107d60, tail = 0x7fffe0107d60}, {head = 0x0, tail = 0x0} <repeats 625 times>}, inpos = 0,
  insize = 0, inbuf = 0x0, is_udp = 0, udp_dest = 0x0, is_broadcast = 0, s = {ss_family = 2, __ss_align = 0, __ss_padding = '\000' <repeats 111 times>},
  auto_reconnect = 1, fragments = 0x0, tcp_syncnt = -1, uid = 0, gid = 0, readahead = 0, debug = 0, ifname = "ffeth3b\000\000\000\000\000\000\000\000",
  timeout = 10000}

(gdb) frame 0
#0  0x00007fffeadbd071 in rpc_destroy_context (rpc=0x7fffe0002b50) at init.c:296
296                LIBNFS_LIST_REMOVE(&q->head, pdu);
(gdb) p q
$10 = (struct rpc_queue *) 0xfffffffc00000001
(gdb) p *q
Cannot access memory at address 0xfffffffc00000001
(gdb) p pdu
$11 = (struct rpc_pdu *) 0x7fffe0107d60
(gdb) p *pdu
$12 = {next = 0x0, xid = 3868610386, zdr = {x_op = ZDR_ENCODE, buf = 0x7fffe0006d10 "\200", size = 1052672, pos = 4, mem = 0x0}, written = 124, outdata = {
    size = 124, data = 0x7fffe01086a0 "\200"}, cb = 0x7fffeadc4256 <nfs_fsync_cb>, private_data = 0x7fffe0107e80,
  zdr_decode_fn = 0x7fffeadd8aa7 <zdr_COMMIT3res>, zdr_decode_buf = 0x0, zdr_decode_bufsize = 144}
(gdb) list
291        for (i = 0; i < HASHES; i++) {
292            struct rpc_queue *q = &rpc->waitpdu[i];
293
294            while((pdu = q->head) != NULL) {
295                pdu->cb(rpc, RPC_STATUS_CANCEL, NULL, pdu->private_data);
296                LIBNFS_LIST_REMOVE(&q->head, pdu);
297                rpc_free_pdu(rpc, pdu);
298            }
299        }
300
(gdb) p i
$1 = 398
(gdb) p rpc->waitpdu[i]
$2 = {head = 0x7fffe0107d60, tail = 0x7fffe0107d60}
(gdb) p &rpc->waitpdu[i]
$3 = (struct rpc_queue *) 0x7fffe0004508
(gdb) p q
$4 = (struct rpc_queue *) 0xfffffffc00000001
(gdb) p pdu->cb
$5 = (void (*)(struct rpc_context *, int, void *, void *)) 0x7fffeadc4256 <nfs_fsync_cb>
```

**Comment by sahlberg** [comment]
Thanks for the report. I will try to look at this on friday, before next weeks release.

About your timeout changes you mention.  I have requests to add such and were planning to add something similar to what I have in libiscsi.
Do you have something that you can send a pull request for for timeout handling?
If not I will build something similar to libiscsi timeouts after the next release.

Thanks!

**Comment by doktorstick** [comment]
I meant to comment earlier RE: timeouts. I went through the effort at the start of this week moving my branch to upstream/master and merging. It turns out we have similar timeout functionality in `wait_for*`. It differs in that you `poll` for the full timeout and `poll` for a maximum of 1 second at a time (psuedo-code):
```c
        while (!cb_data->is_finished) {
                if (now() - t_start > timeout) {
                        rpc_set_error (rpc, "Reply timeout\n");
                        cb_data->status = -EIO;
                        break;
                }
                ...        
```

Still haven't figured out the main topic of this bug, though.

**Comment by sahlberg** [comment]
So it crached during nfs_destroy_context() which does not look like it
should be possible, unless we have either list corruption
or if pdu->cb() also is modifying this list.

Can you reproduce or was this only a single crash ?
(I can not reproduce but will try again)

If you can reproduce, can you try switching the order of these two lines
(two places) in nfs_destroy_context():

pdu->cb(rpc, RPC_STATUS_CANCEL, NULL, pdu->private_data);
LIBNFS_LIST_REMOVE(&rpc->outqueue.head, pdu);

 change into

LIBNFS_LIST_REMOVE(&rpc->outqueue.head, pdu);
pdu->cb(rpc, RPC_STATUS_CANCEL, NULL, pdu->private_data);

**Comment by doktorstick** [comment]
No effect.

Currently I'm updating my code with the latest libnfs to see if there's something that has been changed that makes this problem go away. Looking through the commits, there are some things that are tangentially related that fixes it.

**Comment by earlchew** [comment]
This might be related to https://github.com/sahlberg/libnfs/issues/215

**Comment by doktorstick** [comment]
I've been able to reproduce a crash. The way I do it is by using a modified `nfs-cp` (see the patch below). What I do is create a `nfs-cp` loop as such:
```
client$ echo "Now is the time for all good men..." > file.0
client$ while true; do sudo ./nfs-cp /tmp/file.0 'nfs://172.16.21.99/exports/share2/test.file?if=ffeth1b'; done
copied 36 bytes
copied 36 bytes
...
```
Then, on the filer I create a loop that cycles the interface:
```
filer$ while true; do sudo ifdown eth3; sleep 2; sudo ifup eth3; sleep 2; done
```

The pattern of the `nfs-cp` becomes such--it usually takes a minute or so for it to crash:
```
copied 36 bytes
[... snip ...]
copied 36 bytes
Failed to mount nfs share : Timed out after [1000] milliseconds
Failed to open nfs://172.16.21.99/exports/share2/test.file?if=ffeth1b
Failed to mount nfs share : Timed out after [1000] milliseconds
Failed to open nfs://172.16.21.99/exports/share2/test.file?if=ffeth1b
Failed to mount nfs share : Timed out after [1000] milliseconds
Failed to open nfs://172.16.21.99/exports/share2/test.file?if=ffeth1b
copied 36 bytes
[... snip ...]
copied 36 bytes
Failed to creat file /test.file: Timed out after [1000] milliseconds
Caught SIGSEGV... attaching debugger
GNU gdb (Ubuntu 7.7.1-0ubuntu5~14.04.2) 7.7.1
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
Attaching to process 22754
Reading symbols from /home/chrisr/local/bin/nfs-cp...done.
Reading symbols from /home/chrisr/local/lib/libnfs.so.8...done.
Loaded symbols for /home/chrisr/local/lib/libnfs.so.8
Reading symbols from /lib/x86_64-linux-gnu/libc.so.6...(no debugging symbols found)...done.
Loaded symbols for /lib/x86_64-linux-gnu/libc.so.6
Reading symbols from /lib64/ld-linux-x86-64.so.2...(no debugging symbols found)...done.
Loaded symbols for /lib64/ld-linux-x86-64.so.2
Reading symbols from /lib/x86_64-linux-gnu/libnss_files.so.2...(no debugging symbols found)...done.
Loaded symbols for /lib/x86_64-linux-gnu/libnss_files.so.2
0x00007ffff788554c in waitpid () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) where
#0  0x00007ffff788554c in waitpid () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007ffff7807232 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x000000000040179b in sighandler (signum=11) at nfs-cp.c:226
#3  <signal handler called>
#4  nfs_destroy_context (nfs=<error reading variable: Cannot access memory at address 0xfffffffbffffffd9>,
    nfs@entry=<error reading variable: Cannot access memory at address 0xfffffffc00000009>) at libnfs.c:522
Cannot access memory at address 0xfffffffc00000009
(gdb)
```

Another time I had it:
```
(gdb) where
#0  0x00007ffff0a70b72 in libnfs_zdr_destroy (zdrs=0xffffffffffffffe1) at libnfs-zdr.c:99
#1  0x00007ffff0a7330e in rpc_process_pdu (
    rpc=<error reading variable: Cannot access memory at address 0xffffffffffffff89>,
    buf=<error reading variable: Cannot access memory at address 0xffffffffffffff81>,
    size=<error reading variable: Cannot access memory at address 0xffffffffffffff7d>) at pdu.c:556
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
```

Here's the patch I made to `nfs-cp`. The crux to get this to happen was set the timeout.
```
diff --git a/utils/nfs-cp.c b/utils/nfs-cp.c
index 1af4b74..6b85e5f 100644
--- a/utils/nfs-cp.c
+++ b/utils/nfs-cp.c
@@ -48,6 +48,8 @@ WSADATA wsaData;
 #include <stdlib.h>
 #include <stdint.h>
 #include <sys/types.h>
+#include <signal.h>
+#include <unistd.h>
 #include <fcntl.h>
 #include "libnfs.h"
 #include "libnfs-raw.h"
@@ -172,6 +174,8 @@ open_file(const char *url, int flags)
 		return NULL;
 	}

+        nfs_set_timeout (file_context->nfs, 1000);
+
 	file_context->url = nfs_parse_url_full(file_context->nfs, url);
 	if (file_context->url == NULL) {
 		fprintf(stderr, "%s\n", nfs_get_error(file_context->nfs));
@@ -212,6 +216,18 @@ open_file(const char *url, int flags)
 #define BUFSIZE 1024*1024
 static char buf[BUFSIZE];

+void sighandler (int signum)
+{
+  printf ("Caught SIGSEGV... attaching debugger\n");
+  signal (signum, SIG_DFL);
+
+  char cmd[128];
+  snprintf (cmd, sizeof (cmd)-1,"gdb -p %d", getpid());
+  system (cmd);
+  exit (1);
+}
+
+
 int main(int argc, char *argv[])
 {
 	struct stat st;
@@ -231,6 +247,8 @@ int main(int argc, char *argv[])
 	aros_init_socket();
 #endif

+        signal (SIGSEGV, sighandler);
+
 	if (argc != 3) {
 		usage();
 	}
```

**Comment by earlchew** [comment]
The issue might become obvious if you are able to run the client code under valgrind.

**Comment by doktorstick** [comment]
:)

I had tried a few methods with little success. I had been running with electric fence, but it doesn't report anything (which I expected since the stack is whacked). Compiling `nfs-cp` with ASAN reports:
```
ASAN:SIGSEGV
=================================================================
==15151== ERROR: AddressSanitizer: SEGV on unknown address 0xfffffffbffffffd9 (pc 0x7ffff7b9c37e sp 0x7fffffffe9f0 bp 0xfffffffc00000001 T0)
AddressSanitizer can not provide additional info.
```

And your suggestion of valgrind:
```
Failed to write to dest file
==18076== Invalid read of size 8
==18076==    at 0x4E4A37E: nfs_destroy_context (libnfs.c:522)
==18076==    by 0xFFFFFFFC00000000: ???
==18076==    by 0x54500FF: ???
==18076==    by 0x54500FF: ???
==18076==    by 0x52021FF: ??? (in /lib/x86_64-linux-gnu/libc-2.19.so)
==18076==  Address 0xfffffffbffffffd9 is not stack'd, malloc'd or (recently) free'd
```

**Comment by sahlberg** [comment]
There is also http://rr-project.org/  that could be used to trace back to
where, hopefully, the stack is smashed.

I will try to reproduce tomorrow. And this looks like a good opportunity to
get familiar with rr-project

**Comment by doktorstick** [comment]
After many trials and tribulations of trying to get `rr` to work on a custom 3.16 kernel on a VMware, I went to bare metal hardware with a 4.4 kernel and got it to work. I've not tracked down what is causing the corruption, but I see it happening. I wanted to get this posted before I diving deeper to find out why in the `pdu` structure (I think) things go awry.

You'll observe that the `private_data` in `create_cb` is a bit off.
```
(rr) p *(struct sync_cb_data*) private_data
$49 = {is_finished = 35475568, status = 0, offset = 577, return_data = 0x7ffefd35a740,
  return_int = -566557813, call = 0x21d5040 "\001"}
```

The code is in `create_cb`. It hasn't started modifying values yet. Observe that in frame 3, the `nfs` pointer is correct.
```
(rr) where
#0  creat_cb (status=-4, nfs=0x21d5070, data=0x7fc8de3c626b, private_data=0x7ffefd35a490)
    at libnfs-sync.c:748
#1  0x00007fc8de3abda6 in nfs_create_1_cb (rpc=<optimized out>, status=<optimized out>,
    command_data=0x0, private_data=0x21d9450) at libnfs.c:3352
#2  0x00007fc8de3a9658 in rpc_destroy_context (rpc=0x21d50e0) at init.c:350
#3  0x00007fc8de3b038b in nfs_destroy_context (nfs=0x21d5070) at libnfs.c:543
#4  0x00000000004012e1 in free_file_context (file_context=0x21d5040) at nfs-cp.c:86
#5  0x0000000000401428 in open_file (url=<optimized out>, flags=flags@entry=577) at nfs-cp.c:211
#6  0x0000000000400e86 in main (argc=3, argv=0x7ffefd35a748) at nfs-cp.c:266
```

Now we step forward to the first assignment and observe the call stack; `nfs` gets corrupted (set to `0x1`).
```
(rr) n
745             cb_data->is_finished = 1;
(rr) where
#0  creat_cb (status=-4, nfs=0x21d5070, data=0x7fc8de3c626b, private_data=0x7ffefd35a490)
    at libnfs-sync.c:746
#1  0x00007fc8de3abda6 in nfs_create_1_cb (rpc=<optimized out>, status=<optimized out>,
    command_data=0x0, private_data=0x21d9450) at libnfs.c:3352
#2  0x00007fc8de3a9658 in rpc_destroy_context (rpc=0x21d50e0) at init.c:350
#3  0x00007fc8de3b038b in nfs_destroy_context (nfs=0x1) at libnfs.c:543
#4  0x00000000004012e1 in free_file_context (file_context=0x21d5040) at nfs-cp.c:86
#5  0x0000000000401428 in open_file (url=<optimized out>, flags=flags@entry=577) at nfs-cp.c:211
#6  0x0000000000400e86 in main (argc=3, argv=0x7ffefd35a748) at nfs-cp.c:266
(rr) p *cb_data
$50 = {is_finished = 1, status = 0, offset = 577, return_data = 0x7ffefd35a740, return_int = -566557813, call = 0x21d5040 "\001"}
```

It's further corrupted once `status` is updated and it matches what I've seen... only in `create` path instead of `fsync` path:
```
#3  0x00007fc8de3b038b in nfs_destroy_context (nfs=0xfffffffc00000001) at libnfs.c:543
```

**Comment by doktorstick** [comment]
The crux of the problem is that the local variable `cb_data` in the `*_sync` routines is squirreled away in the `pdu->private_data`. After `wait_for_reply` times out and leaves scope with the timeout error, the `cb_data` ends up getting used as the context is being destroyed. 

Generically, `rpc_nfs3_lookup_async` passes in the next callback to invoke along with the private data. During the `destroy` process, that callback gets invoked[1] and writes the now-invalid stack variable. More generically, this will happen with any of the chaining callbacks that get interrupted in the `*_sync` routines. They'll have the next callback stored in the pdu and `pdu->cb()` during destroy and hit the out-of-scope `cb_data`. I don't know if there are other stack-based `private_data`s in other routines.

I'm don't have a grasp of the implications of not calling `pdu->cb()` in `rpc_error_all_pdus` and `rpc_destroy_context` especially for long-running programs. Commenting out the calls, however, does prevent the crashing. (It would crash within two minutes on my loop test. It's run for about 15 minutes thusfar.)

[1] Since the contexts are being destroyed, why is the next callback being called?

**Comment by sahlberg** [comment]
Thanks,
I have started using rr too and it is awesome.
I did find and fix a different (not the issue above) issue that was related to timeouts in b7fd060bc6b525e6f59a92d9b0ab5f9b3f273368

I will have a look at the thing you found shortly. Thanks!

**Comment by sahlberg** [comment]
I have addressed the SEGV that happened during timeouts.
In libnfs.c there were a lot of places where we only handled RPC_STATUS_ERROR and RPC_STATUS_CANCEL but not RPC_STATUS_TIMEOUT.

This triggered many places where we would encounter an error but the callback would assume everything was ok and try to dereference some pointers that were no longer valid.

I think I have fixed all occurences of that in current master.
